File: src/core/Persister.ts
import { Server } from "./model";
import { LogManager } from "./Logger";

const logger = LogManager.get("core/persistence/Persister");

export abstract class Persister {
    protected persist_for_server(server: Server, key: string, data: any): void {
        this.persist(this.server_key(server, key), data);
    }

    protected persist(key: string, data: any): void {
        try {
            localStorage.setItem(key, JSON.stringify(data));
        } catch (e) {
            logger.error(`Couldn't persist ${key}.`, e);
        }
    }

    protected async load_for_server<T>(server: Server, key: string): Promise<T | undefined> {
        return this.load(this.server_key(server, key));
    }

    protected async load<T>(key: string): Promise<T | undefined> {
        try {
            const json = localStorage.getItem(key);
            return json && JSON.parse(json);
        } catch (e) {
            logger.error(`Couldn't load ${key}.`, e);
            return undefined;
        }
    }

    private server_key(server: Server, key: string): string {
        let k = key + ".";

        // Do this manually per server type instead of just appending e.g. `Server[server]` to
        // ensure the persisted key never changes.
        switch (server) {
            case Server.Ephinea:
                k += "Ephinea";
                break;
            default:
                throw new Error(`Server ${Server[server]} not supported.`);
        }

        return k;
    }
}

File: web/src/main/kotlin/world/phantasmal/web/core/persistence/Persister.kt
package world.phantasmal.web.core.persistence

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import mu.KotlinLogging
import world.phantasmal.web.core.models.Server

private val logger = KotlinLogging.logger {}

abstract class Persister(private val store: KeyValueStore) {
    private val format = Json {
        classDiscriminator = "#type"
        ignoreUnknownKeys = true
    }

    protected suspend inline fun <reified T> persist(key: String, data: T) {
        persist(key, data, serializer())
    }

    // Method suspends so we can use async storage in the future.
    @Suppress("RedundantSuspendModifier")
    protected suspend fun <T> persist(key: String, data: T, serializer: KSerializer<T>) {
        withContext(Dispatchers.Default) {
            logger.trace { """Persisting data with key "$key".""" }

            try {
                store.put(key, format.encodeToString(serializer, data))
            } catch (e: Throwable) {
                logger.error(e) { """Couldn't persist data with key "$key".""" }
            }
        }
    }

    protected suspend inline fun <reified T> persistForServer(
        server: Server,
        key: String,
        data: T,
    ) {
        persist(serverKey(server, key), data)
    }

    protected suspend inline fun <reified T> load(key: String): T? =
        load(key, serializer())

    // Method suspends so we can use async storage in the future.
    @Suppress("RedundantSuspendModifier")
    protected suspend fun <T> load(key: String, serializer: KSerializer<T>): T? =
        withContext(Dispatchers.Default) {
            logger.trace { """Loading persisted data with key "$key".""" }

            try {
                val json = store.get(key)

                if (json == null) {
                    logger.trace { """No persisted data with key "$key".""" }
                    null
                } else {
                    logger.trace { """Loaded persisted data with key "$key".""" }
                    format.decodeFromString(serializer, json)
                }
            } catch (e: Throwable) {
                logger.error(e) { """Couldn't load persisted data with key "$key".""" }
                null
            }
        }

    protected suspend inline fun <reified T> loadForServer(server: Server, key: String): T? =
        load(serverKey(server, key))

    fun serverKey(server: Server, key: String): String {
        // Do this manually per server type instead of just appending e.g. `server` to ensure the
        // persisted key never changes.
        val serverKey = when (server) {
            Server.Ephinea -> "Ephinea"
        }

        return "$key.$serverKey"
    }
}
